Framework ReferenceResidualProblem and auto scaling

lindsayad opened the pull request:
- Moves `ReferenceResidualProblem` to framework
- Makes `ReferenceResidualProblem` work with automatic scaling

lindsayad committed:
Ensure scaling factors done right for Crank-Nicolson

Variable scaling factors must be calculated before the residual
compute call done in CrankNicolson::init

lindsayad committed:
initial -> scaling

lindsayad committed:
Don't use scaling Jacobian for true Jacobian in explicit Euler

lindsayad committed:
Issue warnings in cases where auto scaling cant be conducted

Some cases:
  - PETSc is not available
  - The user is using a non-PETSc matrix
  - There is no matrix but the user has explicitly requested auto
    scaling

lindsayad committed:
Refactor where scaling data lives and is computed from

This commit moves calculation of the scaling Jacobian from the
beginning of NonlinearSystem::solve to PicardSolve::solve. This
ensures that the Picard norm will match the initial non-linear
residual norm of the master application. Additionally all scaling
data members were moved from FEProblemBase and NonlinearSystem
into NonlinearSystemBase. It just seems appropriate that the
nonlinear system (where traditional nonlinear or eigen in the future)
owns the data related to scaling of the residual and Jacobian.
FEProblemBase to me is an object that should be designed to hold
the data related to using finite elements to create an algebraic
system. The nonlinear system then solves that system using Newton
technology (typically)

lindsayad committed:
Move computingScalingJacobian information to SystemBase

This enables:
- Easily setting whether we are computing the scaling jacobian for
  the displaced nonlinear system
- Checking in a kernel whether we are computing the scaling jacobian
  regardless of whether we are in the regular or displaced nonlinear
  system

lindsayad committed:
Dont scale anytime we have JFNK

lindsayad committed:
Remove simple_diffusion test from contact module

lindsayad committed:
Move ReferenceResidualProblem to framework

Closes #9151

lindsayad committed:
ReferenceResidual framework test

lindsayad committed:
Poorly scaled test

lindsayad committed:
RefResidProblem automatic scaling test

lindsayad committed:
Combine RefResidProblem and auto scaling refactor changes

tophmatthews commented:
Dude... As always... 🤘 
Oh and get this in tomorrow already.

tophmatthews commented:
So should I not use `ReferenceResidualProblem` with automatic scaling until this is in? I was with limited success...

lindsayad commented:
You should not. Currently the scaling factors are saved in
ReferenceResidualProblem during initialSetup which is before automatic
scaling has been executed.

On Thu, Sep 5, 2019 at 3:22 PM Topher Matthews <notifications@github.com>
wrote:

> So should I not use ReferenceResidualProblem with automatic scaling until
> this is in? I was with limited success...
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/14004?email_source=notifications&email_token=ACOGA4AM7ORXP2CSWVGI5PDQIGBCPA5CNFSM4IUCEGPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6BAEBQ#issuecomment-528613894>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACOGA4HTLPZN2JPKO26X4DTQIGBCPANCNFSM4IUCEGPA>
> .
>

tophmatthews commented:
oh boy...

tophmatthews commented on modules/contact/test/tests/kernels/simple_diffusion/simple_diffusion.i:
```diff
@@ -1,50 +0,0 @@
-[Mesh]
```
😆

rwcarlsen replied:
@lindsayad - wat?

lindsayad committed:
Convert ReferenceResidualProblem to use tagging

@fdkong has created such a beautiful system :-)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/393578/) on aaaa2de wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14004/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14004/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 03ef85cfdeaf76aab750712eddf6b2e130744cc5`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/393596/) on 61b7700 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14004/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14004/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 03ef85cfdeaf76aab750712eddf6b2e130744cc5`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/393647/) on bfe41a9 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14004/site)

This comment will be updated on new commits.

lindsayad committed:
Make computeScalingJacobian virtual

lindsayad committed:
Change no ref quantity from error to info

There are some augmented lagrange contact tests in the combined module
that do not set reference quantities and rely on an absolute tolerance
to indicate convergence. This seems rather poor, but I'm not going
to go change a bunch of module tests

tophmatthews commented on framework/src/executioners/Executioner.C:
```diff
@@ -57,10 +57,26 @@ Executioner::Executioner(const InputParameters & parameters)
+  if ((_pars.isParamSetByUser("automatic_scaling") && getParam<bool>("automatic_scaling")) &&
+      _fe_problem.solverParams()._type == Moose::ST_JFNK)
+  {
+    mooseWarning("Automatic scaling isn't implemented for the case where you do not have a "
```
paramWarning? Is there even a paramWarning?

tophmatthews commented on framework/src/systems/NonlinearSystem.C:
```diff
@@ -410,3 +396,126 @@ NonlinearSystem::converged()
+    petsc_matrix.reset_preallocation();
+
+#else
+    mooseWarning("Automatic scaling requires a PETSc version of 3.9.0 or greater, so no automatic "
```
why not just error out?

lindsayad replied:
If that is what an application developer would prefer

permcody commented on test/tests/problems/reference_residual_problem/tests:
```diff
@@ -0,0 +1,26 @@
+    type = 'Exodiff'
+    input = 'reference_residual.i'
+    exodiff = 'reference_residual_out.e'
+    requirement = 'ReferenceResidualProblem shall base convergence on comparison of individual variables with reference quantities'
```
```suggestion
    requirement = 'The system shall have the ability to to base base convergence on the comparison of individual variables with with respect to reference quantities of those variables.'
```

lindsayad replied:
Isn't the number one requirement for a requirement that it be unambiguous? "The system" seems so ambiguous! What system? I thought my language was very clear and concise and described what we require `ReferenceResidualProblem` to do, e.g. what I am testing for it to do

permcody replied:
> Isn't the number one requirement for a requirement that it be unambiguous?

Yes, However in traditional development, requirements are written first, sometimes by project managers or other people with less software development experience. They lay down a set of rules or requirements that a new software "system" should satisfy. Arguable, the problem with a long set of requirements is that there is no context to them so there's a lot of repetition in setting them up. It's not typical to say - ok we are working down inside this aspect of the system so let's create a sub block of requirements that roll up to just this section. I've been on projects with hundreds of hand-written requirements that software engineers are supposed to take and run with. It's difficult.

However, from an auditing point of view. The lack of context makes it really easy to single out a requirement, read it gaining all the context from one place and then trace through to the design (where object names and other detail is introduced, not before) and on through to a test that verifies that the requirement is being satisfied.

So at the highest level you can't really do much better than the software "system". Implementation is not supposed to be considered at that level because the software engineer should have the freedom to choose how to implement without over specification. It feels awkward to us, because in practice we usually think about implementation first to achieve a generic goal (perhaps not a formal requirement). Then as we implement we see a lot of corner cases which we implement around creating additional requirements. 

Andrew and I specifically asked about naming objects or even modules (i.e. "The TensorMechanicsModule shall") and we were told not to do it that way because that's getting into implementation details. We were told straight from our NQA-1 consultant and the INL SQA person backed that opinion. A lot of our requirements will need to be overhauled and that will take a lot of time, but I'm trying to prevent new requirements from going in that will need to be revisited.

Hopefully all that background helps. I might need to formalize what I just wrote and add it to the SQA documentation page...
 


 individuals , then software engineers take those

permcody commented on test/tests/problems/reference_residual_problem/tests:
```diff
@@ -0,0 +1,26 @@
+    input = 'reference_residual.i'
+    expect_out = 'Solve Did NOT Converge'
+    cli_args = 'coef=1e17 Outputs/exodus=false -pc_type svd -pc_svd_monitor -ksp_max_it 10'
+    requirement = 'A badly scaled ReferenceResidualProblem will not converge'
```
```suggestion
    requirement = ''The system shall fail to find a converged solution when basing convergence on individual variable reference quantities with poor scaling.''
```

permcody commented on test/tests/problems/reference_residual_problem/tests:
```diff
@@ -0,0 +1,26 @@
+    input = 'reference_residual.i'
+    cli_args = 'coef=1e17 Outputs/file_base=scaled Executioner/automatic_scaling=true Executioner/verbose=true -pc_type svd -pc_svd_monitor -ksp_max_it 10'
+    exodiff = 'scaled.e'
+    requirement = 'An automatically scaled ReferenceResidualProblem will converge'
```
```suggestion
    requirement = 'The system shall have the ability to automatically scale a problem using convergence criteria based on individual reference quantities with poor scaling such that a converged solution may be found.'
```

permcody commented on test/tests/problems/reference_residual_problem/tests:
```diff
@@ -0,0 +1,26 @@
+[Tests]
```
Your requirements still aren't very good at meeting IEEE recommendations. Try to remove the implementation detail when writing the requirement. Assume that the requirement came first so object names and those types of things aren't assumed. This makes it more difficult as your requirement becomes more wordy to try to setup up all of the criteria for making the test unique, but that's precisely the point.

lindsayad replied:
Ok, avoiding the implementation details makes sense to me 👍 . But I can't get over using "the system" when the number one requirement for a requirement is that it be unambiguous.

lindsayad replied:
Ok, just read your response [here](https://github.com/idaholab/moose/pull/14004#discussion_r321797306). Sounds good!

permcody commented on modules/contact/test/tests/kernels/simple_diffusion/tests:
```diff
@@ -1,7 +0,0 @@
-    input = 'simple_diffusion.i'
-    exodiff = 'simple_diffusion_out.e'
-  [../]
-[]
```
Umm... What's going on here?

lindsayad replied:
Do we really want our apps testing simple diffusion? Simple diffusion is done entirely with framework objects. We test that a zillion times every day, so by default it should work in an application.

permcody replied:
Didn't realize this was in the module. This is fine!

![image](https://user-images.githubusercontent.com/3036960/64461224-cf7b4400-d0b9-11e9-8605-4f9595748676.png)

rwcarlsen replied:
Ah - yeah we don't need this test here.

permcody replied:
Also a "zillion" is a bit of an exaggeration. It's more like a million: https://civet.inl.gov/num_tests/

permcody commented on framework/include/systems/NonlinearSystemBase.h:
```diff
@@ -910,4 +919,12 @@ class NonlinearSystemBase : public SystemBase,
       _displaced_mortar_jacobian_functors;
+
+  /// Whether to automatically scale the variables
+  bool _automatic_scaling;
```
Neither one of these Booleans are initialized in the constructor.

lindsayad committed:
Address code review

- Initialize booleans
- Replace relevant mooseErrors and mooseWarnings with paramErrors
  and paramWarnings
- Add error if somehow scaling was requested but we are not using
  a PetscMatrix. Really I don't know how this would happen

lindsayad committed:
Better requirements

tophmatthews commented on framework/src/systems/NonlinearSystem.C:
```diff
@@ -512,8 +512,8 @@ NonlinearSystem::computeScalingJacobian()
 #else
-    mooseWarning("Automatic scaling requires a PETSc version of 3.9.0 or greater, so no automatic "
-                 "scaling is going to be performed");
+    mooseError("Automatic scaling requires a PETSc version of 3.9.0 or greater, so no automatic "
```
paramError?

tophmatthews replied:
The only reason I like param error over regular error is it makes it very clear for the user exactly where it comes with line numbers.

lindsayad replied:
`NonlinearSystem` is not a `MooseObject` so we cannot do `paramError`. However, I've added a check to the `Executioner` constructor that ***will*** generate a `paramError` if automatic scaling has been requested and we don't have the requisite PETSc version.

rwcarlsen replied:
Not that you need to here, but the actual information for making nice paramError-like messages is stored on the InputParameters object itself.  So you could just follow the pattern of code in the paramError implementation that uses that info to make an error message following the same pattern.

lindsayad replied:
Yea but `NonlinearSystem` doesn't have an `InputParameters` object

tophmatthews commented on framework/src/problems/ReferenceResidualProblem.C:
```diff
@@ -81,9 +81,10 @@ ReferenceResidualProblem::ReferenceResidualProblem(const InputParameters & param
-    mooseError("Either the `reference_residual_variables` or `reference_vector` parameter must be "
-               "specified for `ReferenceResidualProblem`. `reference_residual_variables` is "
-               "deprecated so we recommend using `reference_vector`");
+    mooseInfo("Neither the `reference_residual_variables` nor `reference_vector` parameter is "
```
So the idea is that reference residual problem will be the same as nominal behavior if not variables are set, rather than error out?

lindsayad replied:
If you look at the overall PR, I didn't make any changes here. `reference_residual_variables` has never been a required parameter. There are existing augmented lagrange contact tests that do not set this parameter.

I don't know what you mean by nominal behavior. Do you mean like `FEProblem` convergence criteria? If that's what you mean, then no it is not the same. If the parameter is not set, then like the mooseInfo says you'll be relying on the absolute tolerance. No relative tolerance checks will be performed. So ya I can't error out here without breaking existing tests; moreover, I don't even want to make this a warning because downstream apps may be relying on this (rather bad) behavior and if they have `allow_warnings` in their test root, then I will break their testing.

tophmatthews replied:
Hm, maybe there have been some changes to this I don't get until I get hands on.

tophmatthews commented:
I almost think an action should be used to automatically create the "saved_xxx" auxvariables, and `save_in` to be automatically set by kernels, but that is for another time...

tophmatthews commented:
Looks like you just need to restrict two tests to petsc 3.9.0 to get everything to pass

lindsayad commented:
> I almost think an action should be used to automatically create the "saved_xxx" auxvariables, and `save_in` to be automatically set by kernels, but that is for another time...

I'm deprecating the `save_in` way of doing `ReferenceResidualProblem`. Using the tagging system is ***much*** cleaner, and stands to allow us to remove a lot of duplicated code when we eventually remove `save_in`. With tagging there is simply no reason to use `save_in` anymore.

lindsayad commented:
> Looks like you just need to restrict two tests to petsc 3.9.0 to get everything to pass

Got it. Thank you sir 😄

tophmatthews commented:
Tagging system sounds great! You may need to update the doco at some point

tophmatthews commented on framework/src/problems/ReferenceResidualProblem.C:
```diff
@@ -50,18 +51,40 @@ validParams<ReferenceResidualProblem>()
+
   if (params.isParamValid("reference_residual_variables"))
+  {
+    mooseDeprecated("The save-in method for composing reference residual quantities is deprecated "
```
ah, didn't catch all this before

lindsayad committed:
Update ReferenceResidualProblem documentation

lindsayad commented:
> Tagging system sounds great! You may need to update the doco at some point

Good suggestion. Update doco in https://github.com/idaholab/moose/pull/14004/commits/800d8bf2ee3d08d3bac567a7071dab329b9c087b

rwcarlsen reviewed (commented):
That is some terrible block nesting in initialSetup.

lindsayad commented:
> That is some terrible block nesting in initialSetup.

would you prefer that I create two methods that can be called from `initialSetup`, one that is called if we are using tagging and another that is called if we are using `save_in`?
